### ISSUE TITLE
Block Links, Cards, Clickable Regions, Etc.

### ISSUE THREAD

**Issue by nirjan-dev** [nirjan-dev/learning-to-code#276]
[link to the post](https://adrianroselli.com/2020/02/block-links-cards-clickable-regions-etc.html)

 Whether you call them cards, block links, or some other thing, the construct of making an area of content clickable (tappable, Enter-key-able, voice-activatable, etc.) is not new.

While hit area size is mostly a usability issue, marketers often want a larger click area around their calls to action (CTAs) to make them more prominent. Without thinking about the impact on screen reader users, that can make for extremely verbose controls. For voice users, it can be confusing what to say to select one of these.

There are no earth-shattering techniques here. I am mostly demonstrating how doing it the wrong way can be a problem and how a little planning can make the better way less awful. 

# links

 Perhaps the worst thing you can do for a block link is to wrap everything in the <a href>.

If you fail to make the link to display: block, it will have dead areas around words or chunks of text, which can make them feel like distinct links for a mouse user. If you use an underline on the entire block of text, it can be hard to read, while no underline at all can remove that visual affordance.

Worse, for a screen reader user the entire string is read when tabbing through controls. In the following example, the first link contains the heading, image (without declaring it as an image), and block of text, taking about 25 seconds to read before announcing it as a link. When tabbing, you do not always know the control type until the accessible name is complete. 

![image](https://user-images.githubusercontent.com/29270715/75752533-01abec80-5d51-11ea-9e47-b2a0e6d707f9.png)

 The other two links wrap either the <h3> (second example) or the "Read more…" text (third example). For the third one I use aria-describedby to point to the <h3> for more context on the link, but that is not the only nor best way to do it.

The following example demonstrates arrowing through the content. Since there is a misconception among developers that screen reader users Tab through everything on a page, they often forget to test the primary way screen readers scan a page — by arrowing (or swiping if on a touch device). 

 You can hear how every chunk of text is announced as a link, and it is not clear that they are all the same link. Only a more skilled user is likely to recognize the pattern and guess they all go to the same place without further inspection.

The CSS for the block link is generally straightforward: 

```css
.block a[href]::after {
  content: "";
  display: block;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}
```
Note that this approach prevents selecting the text. 

# Variations
I get asked about two challenges to this approach regularly. Content order and additional controls. 

# Content Order

Because I feel strongly that headings should be used to structure your content (content following a heading contextually belongs to that heading), some developers may not know how to mark up a design where an image visually appears above the heading. After all, if the image is in the card, it probably belongs to the heading and text in that card. 

![image](https://user-images.githubusercontent.com/29270715/75752794-83037f00-5d51-11ea-8f04-a7e30e90ea5f.png)

This is one of the few cases where I advocate using CSS to re-order the content to achieve the visual design. If you use my reading order bookmarklet on this, you will see the image is moved above the heading. 

```css
.reorder {
  display: flex;
  flex-direction: column;
}

.reorder h3 {
  order: 2;
}

.reorder p {
  order: 3;
}
```

If you only set order: 1 on the <img> and none of the other content get an order defined, then the <img> will appear after the other content. 

# Content Order — Updated

As Dannie notes in the comments, I could have used the following CSS instead: 

```css
.reorder {
  display: flex;
  flex-direction: column;
}

.reorder img {
  order: -1;
}
```

# Additional Controls
If you need to have an additional link, or a button, or some other control within the card and the design places it at the bottom or top (or even along a side), you can account for this by leaving space for it. This or similar CSS may do the trick: 

```css
.block.extracta a[href]::after {
  bottom: 3.75em;
}

.block.extracta p a[href]::after {
  content: none;
}
```

In my example, I am basing it on the the current text size (em units), but you may need to get more creative if your content will be unknown to you, translated, localized, or otherwise variable. My example is sloppy because I have such a broad selector for links in my cards, so I have to remove the generated content from the additional link. 
![image](https://user-images.githubusercontent.com/29270715/75752955-d675cd00-5d51-11ea-8198-6b7fc2592187.png)

# Buttons

 Most of what I review above applies to <button>s as well. As more and more web content tries to be more appy (think SPAs), a native button might be used to change settings or preferences. Dashboard pages often include card-like interfaces, and those cards similarly stuff way too much into a <button>.

Unlike a link, which allows nested headings, lists, and so on (regardless of how verbose that is), nesting structured content within a <button> removes that structure. An <h3> inside a <button> is no longer a heading. It is just a piece of text. The following video demonstrates that in action. 

Arrowing through the content of a <button> does not mean that structure will be exposed either. 

 A user familiar with the screen probably is fine with a control that announces only as "Settings". Even if confused, it is a simple (and faster) process to arrow backward or jump to the preceding heading for the needed context.

The styles I used to make block links generally apply here. If for some reason your button has an image or needs to have another control within its visual container, the tactics I used above also generally apply.
